src/main/kotlin/me/peckb/aoc/_2024/calendar/day09/Day09.kt

```diff
@@ -0,0 +1,140 @@
+package me.peckb.aoc._2024.calendar.day09
+
+import javax.inject.Inject
+import me.peckb.aoc.generators.InputGenerator.InputGeneratorFactory
+
+class Day09 @Inject constructor(
+  private val generatorFactory: InputGeneratorFactory,
+) {
+  fun partOne(filename: String) = generatorFactory.forFile(filename).readOne { data ->
+    val memory = fillMemory(data)
+
+    defrag(memory, splitFiles = true)
+
+    checksum(memory)
+  }
+
+  fun partTwo(filename: String) = generatorFactory.forFile(filename).readOne { data ->
+    val memory = fillMemory(data)
+
+    defrag(memory, splitFiles = false)
+
+    checksum(memory)
+  }
+
+  private fun fillMemory(line: String): MutableList<Space> {
+    val memory = mutableListOf<Space>()
+
+    var file = true
+    var id = 0
+
+    line.forEach { char ->
+      val length = char.digitToInt()
+      if (file) {
```

Comment:
could make the whole thing a file?.let{ <if true> } ?: <if false>  ... and after it set `file = !file`?